### DISCUSSION TITLE
Messaging FCM token caching

### DISCUSSION THREAD

**Discussion by grigorigoldman** [invertase/react-native-firebase#6716] [Q&A]
I am trying to understand how `getToken` function works in `react-native-firebase/messaging` module. 

From documentation, it implies that it retrieves the token from Firebase, however, I noticed that even if there is no network connection, I can still get the token. This suggests that perhaps it is stored somewhere on the device but I cannot see this documented anywhere.

I checked Firebase SDK documentation as well but there's no mention of storing the token on the device there either.  In fact, the opposite is described. For example, [retrieveFCMToken(forSenderID:)](https://firebase.google.com/docs/reference/swift/firebasemessaging/api/reference/Classes/Messaging#retrievefcmtokenforsenderid:) says that "This registration token is not cached by FIRMessaging."

How is it that calling `getToken` function without network connect returns a current, valid FCM token?

**Comment by mikehardy** [comment]
It sounds like you are referencing the iOS platform? I don't think I can be authoritative here so I humbly suggest you either ask this on stackoverflow (which gets responses from the upstream firebase developers - it's their most official support channel I think) or on the firebase-ios-sdk repo itself

**Reply by grigorigoldman** [reply]
I managed to track it down. It's buried in the `FIRMessagingTokenStore` module. The sequence of calls looks something like this:

`[FIRMessaging retrieveFCMTokenForSenderID]` calls `[FIRMessagingTokenManager tokenWithAuthorizedEntity]`, which then calls `[FIRMessagingTokenManager cachedTokenInfoWithAuthorizedEntity]`.

If the token does not exist or stale, only then does the call to Firebase is made.

I think the docs are somewhat misleading. Technically speaking, yes `FIRMessaging` does not cache the token but the module it uses does 🤷‍♂️
